spec/controllers/moderator_applications_controller_spec.rb

```diff
@@ -0,0 +1,174 @@
+
+  it 'allows admin access even if applications are closed', :admin_login, :applications_closed do
+    expect_successful_template route
+  end
+
+  it 'works' do
+    expect_successful_template route
+  end
+end
+
+describe ModeratorApplicationsController do
+  before(:each) do |example|
+    ENV['MODERATOR_APPLICATIONS'] = example.metadata[:applications_closed] ? 'closed' : 'open'
+
+    @user = create(:user)
+    setup_session(example, @user)
+
+    @application = build(:application, user: @user)
+    @application.save! unless example.metadata[:does_not_exist]
+  end
+
+  describe 'GET :index' do
+    before(:each) { get :index }
+
+    it_requires_login
+
+    it 'redirects to :new if the user is not an admin and has not yet applied', :does_not_exist do
+      expect(response).to redirect_to(new_moderator_application_path)
+    end
+
+    it 'redirects to :show if the user is not an admin but has applied' do
+      expect(response).to redirect_to(moderator_application_path(@application))
+    end
+
+    it 'works', :admin_login do
+      expect_successful_template :index
+    end
+
+    it 'works even if applications are closed', :admin_login, :applications_closed do
+      expect_successful_template :index
+    end
+  end
+
+  describe 'GET :new' do
+    before(:each) { get :new }
+
+    it_requires_login
+    it_requires_that_applications_are_open
+
+    it 'redirects if the user already applied' do
+      expect(response).to redirect_to(edit_moderator_application_path(@application))
+    end
+
+    it 'works', :does_not_exist do
+      expect_successful_template :new
+    end
+  end
+
+  describe 'POST :create' do
+    before(:each) { post :create, params: { moderator_application: @application.attributes } }
+
+    it_requires_login
+    it_requires_that_applications_are_open
+
+    it 'redirects if the user already applied' do
+      expect(response).to redirect_to(edit_moderator_application_path(@application))
+    end
+
+    it 'works', :does_not_exist do
+      expect(response).to have_http_status(:see_other)
+    end
+  end
+
+  describe 'GET :show' do
+    show_and_edit_tests(:show)
+  end
+
+  describe 'GET :edit' do
+    show_and_edit_tests(:edit)
+  end
+
+  describe 'PUT :update' do
+    before(:each) do
+      put :update, params: { id: application_id, moderator_application: @application.attributes }
+    end
+
+    it_handles_nonexistent_application
+    it_requires_correct_login
+    it_requires_that_applications_are_open
+
+    it 'works' do
+      expect(response)
+        .to redirect_to(moderator_application_path(@application))
+        .and have_http_status(:see_other)
+    end
+  end
+
+  describe 'DELETE :destroy' do
+    before(:each) { delete :destroy, params: { id: application_id } }
+
+    it_handles_nonexistent_application
+    it_requires_correct_login
+    it_requires_that_applications_are_open
+
+    expect_record_missing = lambda do
+      expect(ModeratorApplication.find_by(id: @application.id)).to be_nil
```

Comment:
Do you suppose it would be too much of a hack to turn this into:
```ruby
expect_missing @application
```
which dynamically dispatches on the class to `find_by` the `id` and perform the nil expectation?

It *is* Ruby after all. Ruby, like Smalltalk, seems to be designed to allow you to write whatever the fuck abstractions you want. It'd probably be considered bad to do if this were Java, but I think in Smalltalk it might be normal. Not sure where Ruby lies in that spectrum. I like the Perl+Smalltalk vibe. Ruby gives me the impression of being a hacker's language, even if it's a relatively shitty one.